8898 openssl 110

markrwilliams opened the pull request:
Based on the patch in http://twistedmatrix.com/trac/ticket/8898

markrwilliams committed:
OpenSSL >= 1.1.0 requires keys be at least 4096 bytes.

markrwilliams committed:
OpenSSL >= 1.1.0 removes SSLv2 entirely.

SSL.OP_NO_SSLv2 is now 0, so while it's still safe to set, tests
cannot assert that it was.

markrwilliams committed:
OpenSSL >= 1.1.0 fails with EOF when a client cannot complete a handshake.

Terminating a TLS connection after the client sends a ClientHello but
before it completes a handshake always fails.  The tests in this
commit terminate client connections in this way.

Earlier versions of OpenSSL fail the BIO flush with "SSL routines: ssl
handshake failure" error in ssl3_read_bytes.  The affected tests
expected this by only trapping Error.

The TLS state machine rewrite in 1.1.0 (see
https://www.openssl.org/news/openssl-1.1.0-notes.html) presumably
resulted in the BIO flush failing with EOF.

markrwilliams commented:
[`cryptography` 1.8's macOS and Windows wheels now statically link against OpenSSL 1.1.0](https://cryptography.io/en/latest/changelog/#id2), so this PR should only address test failures on those platforms.

Lukasa commented on src/twisted/test/test_ssl.py:
```diff
@@ -667,8 +671,10 @@ def test_method(self):
-        # And OP_NO_SSLv2 disables the SSLv2 support.
-        self.assertTrue(self.context._options & SSL.OP_NO_SSLv2)
+        # OpenSSL >= 1.1.0 no longer provides SSLv2
+        if not OPENSSL_1_1:
```
This is not really the right question. Earlier versions of OpenSSL could have SSLv2 compiled out too. You should probably look for the option flag, not the OpenSSL version.

markrwilliams replied:
> Earlier versions of OpenSSL could have SSLv2 compiled out too.

True, but [`SSL_OP_NO_SSLv2` is defined as zero as of 1.1.0](https://github.com/openssl/openssl/blob/OpenSSL_1_1_0/include/openssl/ssl.h#L339).  So it's fine to set it (`flags | 0` -> `flags`) but impossible to check (`flags & 0` -> `0`).

markrwilliams replied:
Sorry, that was a bit terse.  You can compile prior versions of OpenSSL without SSLv2 support, but `SSL_OP_NO_SSLv2` will continue to be defined as some non-zero value.  [In 1.0.2ish, for example, there are no `ifdef`s around its definition](https://github.com/openssl/openssl/blob/OpenSSL_1_0_2-stable/ssl/ssl.h#L644). 

That means this test could pass even on versions that lacked SSLv2 support, because it was possible to set the `SSL_OP_NO_SSLv2` flag and recover it from the context's options.

In 1.1.0+, however, `SSL_OP_NO_SSLv2` is 0.  That means it can be set but it cannot even be checked because it has no effect!

That's why I put the OpenSSL version check here.  We _definitely_ want to ensure we're excluding SSLv2 on earlier versions of OpenSSL where it's even at all possible.

Lukasa replied:
If it's present but set to zero, why do we care? Just have the test say `if OP_NO_SSLv2` before it asserts the flag is set.

alex replied:
You can still test the flag when the value is 0! `assert (value & SSL_OP_NO_SSLv2) == SSL_OP_NO_SSLv2`

markrwilliams replied:
Hm, I thought it was clearer to call out the OpenSSL version in code, but a comment will do.  I'll switch the asserts to check that the value's in the options as @alex suggested.

Lukasa commented on src/twisted/test/test_ssl.py:
```diff
@@ -714,7 +720,8 @@ def test_method(self):
         """
         self.assertEqual(self.context._method, SSL.SSLv23_METHOD)
-        self.assertTrue(self.context._options & SSL.OP_NO_SSLv2)
+        if not OPENSSL_1_1:
```
Same here.

Lukasa reviewed (changes requested):
Generally looks good, one minor note.

alex commented on src/twisted/test/server.pem:
```diff
@@ -6,7 +6,7 @@ from datetime import datetime
 
 key = PKey()
-key.generate_key(TYPE_RSA, 2048)
+key.generate_key(TYPE_RSA, 4096)
```
Why is this changing?

markrwilliams replied:
It was in the patch I pulled out of 8898.

alex replied:
2048 bit keys are fine, so I'd revert unrelated changes like this.

alex commented:
It is _not true_ that keys must be 4096 bits, 2048 bit keys are fine.

markrwilliams commented:
@alex The only key size failure appears to be in the public keys for X509 end-entity certificates.  Is 2048 bits the minimum key size for those?  Any chance you have a citation I can put in a comment or commit message?

alex commented:
A citation for OpenSSL _not_ making a change?

On Mon, Mar 13, 2017 at 10:44 AM, Mark Williams <notifications@github.com>
wrote:

> @alex <https://github.com/alex> The only key size failure appears to be
> in the public keys for X509 end-entity certificates. Is 2048 bits the
> minimum key size for those? Any chance you have a citation I can put in a
> comment or commit message?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/twisted/twisted/pull/743#issuecomment-286128245>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAADBOvMTLtdYkZqTOckeCznJj09ERyDks5rlVZFgaJpZM4Mafgs>
> .
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: D1B3 ADC0 E023 8CA6

markrwilliams commented:
@alex This is the citation I was looking for:

> The functions SSL_CTX_set_security_level() and SSL_set_security_level() set the security level to level. If not set the library default security level is used. [...]
> Level 1
The security level corresponds to a minimum of 80 bits of security. Any parameters offering below 80 bits of security are excluded. **As a result RSA, DSA and DH keys shorter than 1024 bits and ECC keys shorter than 160 bits are prohibited.** All export ciphersuites are prohibited since they all offer less than 80 bits of security. SSL version 2 is prohibited. Any ciphersuite using MD5 for the MAC is also prohibited. [...]
> The default security level can be configured when OpenSSL is compiled by setting -DOPENSSL_TLS_SECURITY_LEVEL=level. **If not set then 1 is used**. [...]

https://www.openssl.org/docs/man1.1.0/ssl/SSL_CTX_get_security_level.html

IOW, OpenSSl 1.1.0 introduced "security levels" to SSLContext, and the default (1) requires RSA keys be at least 1024 bits.  Does that look right?  If so I'll put a link to that in the relevant commit.

alex commented:
Yeah, that looks correct.

On Tue, Mar 14, 2017 at 11:46 AM, Mark Williams <notifications@github.com>
wrote:

> @alex <https://github.com/alex> This is the citation I was looking for:
>
> The functions SSL_CTX_set_security_level() and SSL_set_security_level()
> set the security level to level. If not set the library default security
> level is used. [...]
> Level 1
> The security level corresponds to a minimum of 80 bits of security. Any
> parameters offering below 80 bits of security are excluded. *As a result
> RSA, DSA and DH keys shorter than 1024 bits and ECC keys shorter than 160
> bits are prohibited.* All export ciphersuites are prohibited since they
> all offer less than 80 bits of security. SSL version 2 is prohibited. Any
> ciphersuite using MD5 for the MAC is also prohibited. [...]
> The default security level can be configured when OpenSSL is compiled by
> setting -DOPENSSL_TLS_SECURITY_LEVEL=level. *If not set then 1 is used*.
> [...]
>
> https://www.openssl.org/docs/man1.1.0/ssl/SSL_CTX_get_security_level.html
>
> IOW, OpenSSl 1.1.0 introduced "security levels" to SSLContext, and the
> default (1) requires RSA keys be at least 1024 bits. Does that look right?
> If so I'll put a link to that in the relevant commit.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/twisted/twisted/pull/743#issuecomment-286463789>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAADBC2pA3Di-UStpAuIEgi8Z48CO2ucks5rlrZegaJpZM4Mafgs>
> .
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: D1B3 ADC0 E023 8CA6

markrwilliams commented:
The change 9bf53b7 describes was introduced in https://github.com/openssl/openssl/commit/9ab930b27d51a13362e6647074f13589a8ac004d. [This gist](https://gist.github.com/markrwilliams/ff25334d544fcb10cb6231697b518b3b) contains the code used to discover the commit.

The change in behavior is roughly as follows: [`tls_get_message_header` calls `ssl3_read_bytes` but treats a zero-length read and an error the same](https://github.com/openssl/openssl/commit/9ab930b27d51a13362e6647074f13589a8ac004d#diff-c5c4a7b44c0f50fcaec06a0c7cfbd012R498), so that [its caller ` ssl3_get_message` cannot distinguish between the two and itself returns -1](https://github.com/openssl/openssl/commit/9ab930b27d51a13362e6647074f13589a8ac004d#diff-c5c4a7b44c0f50fcaec06a0c7cfbd012R437).  This eventually prevents the handshake function from ever returning 0, and thus [no handshake failure alert occurs](https://github.com/openssl/openssl/blob/9ab930b27d51a13362e6647074f13589a8ac004d/ssl/record/rec_layer_s3.c#L1033-L1034).

[This](https://github.com/openssl/openssl/blob/OpenSSL_1_1_0e/ssl/statem/statem_lib.c#L352-L359) [behavior](https://github.com/openssl/openssl/blob/OpenSSL_1_1_0e/ssl/statem/statem.c#L526-L538) [also exists](https://github.com/openssl/openssl/blob/OpenSSL_1_1_0e/ssl/statem/statem.c#L385-L394) [with](https://github.com/openssl/openssl/blob/OpenSSL_1_1_0e/ssl/statem/statem.c#L168-L171) [the new](https://github.com/openssl/openssl/blob/OpenSSL_1_1_0e/ssl/methods.c#L18-L21) [state](https://github.com/openssl/openssl/blob/OpenSSL_1_1_0e/ssl/ssl_lib.c#L3053) [machine in 1.1.0](https://github.com/openssl/openssl/blob/OpenSSL_1_1_0e/ssl/record/rec_layer_s3.c#L1015-L1024).

So 9bf53b7's commit message is accurate, and I think the implementation is safe enough in that it does not obscure a deeper bug.  Unfortunately I can't rule out _other_ latent issues introduced by the TLS parsing/state machine rewrite in OpenSSL 1.1.0.

This particular issue seems like a bug in OpenSSL, because the handshake failure branch may now be dead code.  I'll investigate there after closing this issue.

markrwilliams committed:
Revert "OpenSSL >= 1.1.0 requires keys be at least 4096 bytes."

This reverts commit e2fa89f2cc2e32a2da76d1d3e37cfd9122b40af5.

markrwilliams committed:
OpenSSL 1.1.0 requires RSA keys be at least 1024 bits.

OpenSSL 1.1.0 introduced "security levels" to SSL Contexts.  The
default security level is 1, and it rquires that RSA keys be at least
1024 bits.

See: https://www.openssl.org/docs/man1.1.0/ssl/SSL_CTX_get_security_level.html

markrwilliams committed:
Revert "OpenSSL >= 1.1.0 removes SSLv2 entirely."

This reverts commit b9621a8559cd4016fa81d6bbb9227306730050d5.

markrwilliams committed:
OpenSSL >= 1.1.0 removes SSLv2 entirely.

SSL.OP_NO_SSLv2 is now 0, so tests cannot assert that it has been
applied by checking its truth value.  Instead they should check that
ANDing SSL.OP_NO_SSLv2 with some flags equals SSL.OP_NO_SSLv2, since N
& X = X when X is 0 or X > 1 and set in N.

markrwilliams committed:
topfile

markrwilliams commented:
@Lukasa This is now officially up for review!

https://twistedmatrix.com/trac/ticket/8898#comment:3

glyph commented:
THANK YOU for doing this everybody. This was blocking a huge amount of other work.

(In related news, we _really_ need to pin our dependencies for CI and use requires.io or something so that these types of fixes can happen without blocking literally everything else...)
